### PR TITLE
RES-116 Changes to temporal pooler monitor mixin and Region support

### PR DESCRIPTION
**Pull request by SaganBolliger**
Addresses https://jira.numenta.com/browse/RES-116

@scottpurdy @chetan51 It would be great to get some feedback on this.

### CODE REVIEW

File: htmresearch/support/union_temporal_pooler_monitor_mixin.py [hunk 1]
```diff
@@ -28,30 +28,180 @@
...
   Mixin for UnionTemporalPooler that stores a detailed history, for inspection and
   debugging.
   """
 
   def __init__(self, *args, **kwargs):
+    if "distanceMetric" in kwargs:
+      self._mmDistanceMetric = kwargs["distanceMetric"]
+      del kwargs["distanceMetric"]
+    else:
+      self._mmDistanceMetric = "pctOverlapLarger"
```
Comment (chetan51) [inline htmresearch/support/union_temporal_pooler_monitor_mixin.py]: You can use `kwargs.get("distanceMetric", "pctOverlapLarger")`.

File: htmresearch/support/union_temporal_pooler_monitor_mixin.py [hunk 2]
```diff
@@ -28,30 +28,180 @@
...
+
     super(UnionTemporalPoolerMonitorMixin, self).__init__(*args, **kwargs)
 
     self._mmResetActive = True  # First iteration is always a reset
+    self.mmClearHistory()
```
Comment (chetan51) [inline htmresearch/support/union_temporal_pooler_monitor_mixin.py]: Why is this necessary?

File: htmresearch/support/union_temporal_pooler_monitor_mixin.py [hunk 3]
```diff
@@ -28,30 +28,180 @@
...
+
+
+  def mmGetDataUnionSDRDutyCycle(self):
+    """
+    @return (list) duty cycles for union SDR bits
+    """
+    return self._mmData["UnionSDRDutyCycle"].tolist()
```
Comment (chetan51) [inline htmresearch/support/union_temporal_pooler_monitor_mixin.py]: Naming convention: `UnionSDRDutyCycle` should be  `unionSDRDutyCycle`.

File: htmresearch/support/union_temporal_pooler_monitor_mixin.py [hunk 4]
```diff
@@ -279,22 +544,28 @@ def compute(self, *args, **kwargs):
...
+    
+    unionSDR = UnionTemporalPooler.compute(self, *args, **kwargs)
+    
+    ### From spatial pooler
+    # total number of connections
+    connectedCounts = numpy.zeros(self.getNumColumns(), dtype=uintType)
+    self.getConnectedCounts(connectedCounts)
+    numConnections = numpy.sum(connectedCounts)
+
+    # import pdb; pdb.set_trace()
```
Comment (chetan51) [inline htmresearch/support/union_temporal_pooler_monitor_mixin.py]: Remove commented-out code

File: htmresearch/support/union_temporal_pooler_monitor_mixin.py (hunk 2 above)
Comment (SaganBolliger) [inline htmresearch/support/union_temporal_pooler_monitor_mixin.py]: mmClearHistory() initializes the arrays in _mmData, so instead of initializing them separately in __init__(), calling mmClearHistory() from __init__() minimizes repetition.

File: htmresearch/support/union_temporal_pooler_monitor_mixin.py (hunk 1 above)
Comment (SaganBolliger) [inline htmresearch/support/union_temporal_pooler_monitor_mixin.py]: Thanks. Actually looks like `kwargs.pop("distanceMetric", "pctOverlapLarger")` would be better since that also has the effect of removing the key from the dictionary.

**Comment by SaganBolliger** [comment]
Thanks @chetan51 for your feedback. I've implemented those changes. I've also switched to using raw overlap scores rather than a normalized distance metric as per @subutai's suggestion.

@BoltzmannBrain: Is there anything that you'd like to see here that isn't already implemented?

File: htmresearch/support/union_temporal_pooler_monitor_mixin.py [hunk 5]
```diff
@@ -45,13 +52,150 @@ def __init__(self, *args, **kwargs):
     super(UnionTemporalPoolerMonitorMixin, self).__init__(*args, **kwargs)
 
     self._mmResetActive = True  # First iteration is always a reset
+    self.mmClearHistory()
```
Comment (chetan51) [inline htmresearch/support/union_temporal_pooler_monitor_mixin.py]: Do you think we should put this in the base `MonitorMixin` class

Comment (SaganBolliger) [inline htmresearch/support/union_temporal_pooler_monitor_mixin.py]: I don't see how it would break anything at least.

File: htmresearch/regions/TemporalPoolerRegion.py [hunk 6]
```diff
@@ -22,16 +22,26 @@
 import numpy
 from nupic.bindings.math import GetNTAReal
 from htmresearch.algorithms.union_temporal_pooler import UnionTemporalPooler
+from htmresearch.support.union_temporal_pooler_monitor_mixin \
+  import UnionTemporalPoolerMonitorMixin
```
Comment (BoltzmannBrain) [inline htmresearch/regions/TemporalPoolerRegion.py]: Please use implicit continuation w/ parenthesis, not backslashes.

**Comment by BoltzmannBrain** [comment]
@SaganBolliger LGTM
